python/mxnet/text/glossary.py

```diff
@@ -0,0 +1,654 @@
+
+        token_cap = len(self._counter) if top_k_freq is None \
+            else len(specials) + top_k_freq
+
+        for token, freq in token_freqs:
+            if freq < min_freq or len(self._idx_to_token) == token_cap:
+                break
+            self._idx_to_token.append(token)
+            self._token_to_idx[token] = len(self._idx_to_token) - 1
+
+    def __len__(self):
+        return len(self.idx_to_token)
+
+    @property
+    def counter(self):
+        return self._counter
+
+    @property
+    def token_to_idx(self):
+        return self._token_to_idx
+
+    @property
+    def idx_to_token(self):
+        return self._idx_to_token
+
+    @property
+    def idx_to_vec(self):
+        return self._idx_to_vec
+
+    @property
+    def vec_len(self):
+        return self._vec_len
+
+    @property
+    def specials(self):
+        return self._specials
+
+    @staticmethod
+    def unk_idx():
+        """The index for unknown tokens (the first token in `specials`)."""
+        return 0
+
+    def set_idx_to_vec(self, embeds):
+        """Sets the mapping between token indices and token embedding vectors.
+
+
+        Parameters
+        ----------
+        embeds : mxnet.text.Embedding or list of mxnet.text.Embedding. If it
+            is a list of mxnet.text.Embedding instances, their embedding vectors
+            are concatenated for each token.
+        """
+
+        # Sanity check.
+        if isinstance(embeds, list):
+            for loaded_embed in embeds:
+                assert isinstance(loaded_embed, Embedding)
+        else:
+            assert isinstance(embeds, Embedding)
+            embeds = [embeds]
+
+        self._vec_len = sum(embed.vec_len for embed in embeds)
+        self._idx_to_vec = nd.zeros(shape=(len(self), self.vec_len))
+
+        col_start = 0
+        # Concatenate all the embedding vectors in embeds.
+        for embed in embeds:
+            col_end = col_start + embed.vec_len
+            self._idx_to_vec[:, col_start:col_end] = embed[self.idx_to_token]
+            col_start = col_end
+
+    # tokens is a str or a list of strs.
+    def update_idx_to_vec(self, tokens, new_vectors):
+        """Updates embedding vectors for tokens.
+
+
+        Parameters
+        ----------
+        tokens : str or a list of strs.
+            A token or a list of tokens whose embedding vector are to be
+            updated.
+        new_vectors : mxnet.ndarray.NDArray
+            A 2-D NDArray to be assigned to the embedding vectors of `tokens`.
+            Its length must be equal to the number of `tokens` and its width
+            must be equal to the dimension of embeddings of the glossary.
+        """
+
+        assert self.idx_to_vec is not None, \
+            'mxnet.text.Glossary._idx_to_vec has not been initialized.  Use ' \
+            'mxnet.text.Glossary.__init__() or ' \
+            'mxnet.text.Glossary.set_idx_to_embed() to initialize it.'
```

Comment:
The assertion message includes internal implementation details. Consider removing such reference and change to something like “Glossary has not been initialized. Do X...”

Reply:
Resolved.